gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/upgrades/NonBCryptIterationsRoundsUpgrader.java

```diff
@@ -0,0 +1,162 @@
+                .flatMap(updatedTask -> {
+                    if (updatedTask.getOperationId().equals(instanceOperationId)) {
+                        return migrate(updatedTask)
+                                .andThen(Single.just(true))
+                                .onErrorResumeNext((err) -> {
+                                    log.error("Unable to upgrade non-BCrypt password encoders (task: {}): {}", TASK_ID, err.getMessage());
+                                    return Single.just(false);
+                                });
+                    } else {
+                        return Single.error(new IllegalStateException("Task " + getTaskId() + " already processed by another instance : trigger a retry"));
+                    }
+                });
+    }
+
+    private Completable migrate(SystemTask task) {
+        return idpService.findAll()
+                .map(idp -> new IdpData(idp, objectMapper))
+                .filter(idp -> !idp.cfg.hasValue("passwordEncoder", "BCrypt"))
+                .flatMapSingle(this::update)
+                .ignoreElements()
+                .doOnError(err -> updateSystemTask(task, (SystemTaskStatus.FAILURE), task.getOperationId()).subscribe())
+                .andThen(updateSystemTask(task, SystemTaskStatus.SUCCESS, task.getOperationId()).ignoreElement());
+    }
+
+    @SneakyThrows
+    private Single<IdentityProvider> update(IdpData idpData) {
+        UpdateIdentityProvider updateObject = createUpdateObject(idpData);
+        return idpService
+                .update(idpData.idp.getReferenceType(), idpData.idp.getReferenceId(), idpData.idp.getId(), updateObject, null, true)
+                .doOnSuccess(updated -> log.info("Removed passwordEncoderOptions for idp={} as a migration process.", updated.getId()));
+    }
+
+    private UpdateIdentityProvider createUpdateObject(IdpData idpData) {
+        var idp = idpData.idp;
+        idpData.cfg.removeProperty("passwordEncoderOptions");
+
+        UpdateIdentityProvider updateIdentityProvider = new UpdateIdentityProvider();
+        updateIdentityProvider.setConfiguration(idpData.cfg.toJson(objectMapper));
+        updateIdentityProvider.setMappers(idp.getMappers());
+        updateIdentityProvider.setName(idp.getName());
+        updateIdentityProvider.setPasswordPolicy(idp.getPasswordPolicy());
+        updateIdentityProvider.setRoleMapper(idp.getRoleMapper());
+        updateIdentityProvider.setDomainWhitelist(idp.getDomainWhitelist());
+        return updateIdentityProvider;
+    }
+
+    @Override
+    protected IllegalStateException getIllegalStateException() {
+        return new IllegalStateException("Couldn't remove iteration rounds for non-BCrypt password encoders");
+
+    }
+
+    @Override
+    protected String getTaskId() {
+        return TASK_ID;
+    }
+
+    @Override
+    public int getOrder() {
+        return DOMAIN_IDP_UPGRADER;
+    }
+
+    @Value
+    private static class IdpData {
```

Comment:
Can't it be a record?

Reply:
the cfg property is 'loaded'. so its still not only data structure.